### PR TITLE
[stateless_validation] Cleanup for bypass sending stateless validation messages to ourselves

### PR DESCRIPTION
**Pull request by shreyan-gupta**
I wasn't too happy exposing chunk_endorsements cache in `ChunkEndorsementTracker` that was introduced as part of this PR: https://github.com/near/nearcore/pull/10531

Hopefully this is better structured. Suggestions are welcomed!

### CODE REVIEW

File: chain/client/src/stateless_validation/state_witness_producer.rs [hunk 1]
```diff
@@ -48,23 +48,22 @@ impl Client {
...
-        // chunk endorsement related network messages.
         if let Some(my_signer) = self.validator_signer.clone() {
-            let validator_id = my_signer.validator_id();
-            if chunk_validators.contains(validator_id) {
-                // Endorse the chunk immediately, bypassing sending state witness
-                // to ourselves, because network can't send messages to ourselves.
-                // Also useful in tests where we don't have a good way to handle
-                // network messages and there's only a single client.
+            // Remove ourselves from the list of chunk validators. Network can't send messages to ourselves.
+            chunk_validators.retain(|validator| {
```
Comment (wacban) [inline chain/client/src/stateless_validation/state_witness_producer.rs:53]: It's a bit strange to do the bypass inside the retain method closure tbh. I would suggest doing retain and bypass separately.

File: chain/client/src/stateless_validation/chunk_validator.rs [hunk 2]
```diff
@@ -60,8 +60,15 @@ impl ChunkValidator {
         epoch_manager: Arc<dyn EpochManagerAdapter>,
         network_sender: Sender<PeerManagerMessageRequest>,
         runtime_adapter: Arc<dyn RuntimeAdapter>,
+        chunk_endorsement_tracker: Arc<ChunkEndorsementTracker>,
```
Comment (wacban) [inline chain/client/src/stateless_validation/chunk_validator.rs:63]: I have a bit of a mixed feelings about this, seem a bit strange to have the validator have access to the tracker. That being said I like it more than what we have now so you have my approval.

Comment (shreyan-gupta) [reply chain/client/src/stateless_validation/chunk_validator.rs:63]: Figuring out which module depends on what is a HUGE headache imo.... I spent a good amount of time thinking about this and I felt this is probably the best approach.

The other idea I had was to create two separate traits, one related to chunk endorsement tracking and one related to validation, and have a single module implement both traits. That way there wouldn't technically be any dependency between these as they are no longer modules, just traits....

File: chain/client/src/stateless_validation/state_witness_producer.rs (hunk 1 above)
Comment (shreyan-gupta) [reply chain/client/src/stateless_validation/state_witness_producer.rs]: Yeah, makes sense...

**Review by pugachAG — APPROVED** [review]
this definitely makes it better!

**Pull request merged**